haystack/utils/device.py

```diff
@@ -1,34 +1,471 @@
+
+        return cls(_single_device=device)
+
+    @classmethod
+    def from_multiple(cls, device_map: DeviceMap) -> "ComponentDevice":
+        """
+        Create a component device representation from a device map.
+
+        :param device_map:
+            The device map.
+        :returns:
+            The component device representation.
+        """
+        return cls(_multiple_devices=device_map)
+
+    def to_torch(self) -> "torch.device":
+        """
+        Convert the component device representation to PyTorch format.
+        Device maps are not supported.
+
+        :returns:
+            The PyTorch device representation.
+        """
+        if self._single_device is None:
+            raise ValueError("Only single devices can be converted to PyTorch format")
+
+        torch_import.check()
+        assert self._single_device is not None
+        return torch.device(str(self._single_device))
+
+    def to_torch_str(self) -> str:
+        """
+        Convert the component device representation to PyTorch string format.
+        Device maps are not supported.
+
+        :returns:
+            The PyTorch device string representation.
+        """
+        if self._single_device is None:
+            raise ValueError("Only single devices can be converted to PyTorch format")
+
+        assert self._single_device is not None
+        return str(self._single_device)
+
+    def to_spacy(self) -> int:
+        """
+        Convert the component device representation to spaCy format.
+        Device maps are not supported.
+
+        :returns:
+            The spaCy device representation.
+        """
+        if self._single_device is None:
+            raise ValueError("Only single devices can be converted to spaCy format")
+
+        assert self._single_device is not None
+        if self._single_device.type == DeviceType.GPU:
+            assert self._single_device.id is not None
+            return self._single_device.id
+        else:
+            return -1
+
+    # pylint: disable=inconsistent-return-statements
+    def to_hf(self) -> Union[Union[int, str], Dict[str, Union[int, str]]]:
+        """
+        Convert the component device representation to HuggingFace format.
+
+        :returns:
+            The HuggingFace device representation.
+        """
+
+        def convert_device(device: Device, *, gpu_id_only: bool = False) -> Union[int, str]:
+            if gpu_id_only and device.type == DeviceType.GPU:
+                assert device.id is not None
+                return device.id
+            else:
+                return str(device)
+
+        if self._single_device is not None:
+            return convert_device(self._single_device)
+        elif self._multiple_devices is not None:
+            return {
+                key: convert_device(device, gpu_id_only=True) for key, device in self._multiple_devices.mapping.items()
+            }
+        else:
+            # Unreachable
+            assert False
```

Comment:
I think we can restructure this to make pylint happy:

```python
        if self._single_device is not None:
            return convert_device(self._single_device)
        return {
            key: convert_device(device, gpu_id_only=True) for key, device in self._multiple_devices.mapping.items()
        }
```

if we really need to check that `self._multiple_devices` is not None before running the second statement, then I'd rather make a preventive check such as:

```suggestion
        if self._single_device is None and self._multiple_devices is None:
            raise ValueError("either _single_device or _multiple_device must have a value")

        if self._single_device is not None:
            return convert_device(self._single_device)
        return {
            key: convert_device(device, gpu_id_only=True) for key, device in self._multiple_devices.mapping.items()
        }
```

(don't forget to delete the pylint comment at the top of the function if you accept this change)

Reply:
Fixed.